### ISSUE TITLE
getentropy on Linux?

### ISSUE THREAD

**Issue by lambdafu** [randombit/botan#1322]
`src/lib/entropy/getentropy/info.txt` seems to suggest getentropy is only used on openbsd. The system call is also available on Linux systems now, and it's obviously a nice improvement over /dev/urandom (which, if I interpret things correctly, is currently used by default on Linux). Are there any plans to enable getentropy for Linux and make it the new default? Thanks!

**Comment by randombit** [comment]
No immediate plans. The main issue is that support was only added in glibc in 2.25, released earlier this year. So it is not available in most distros, basically right now only in rolling release distros like Arch and Gentoo.

And, we cannot really enable it conditionally right now, because of a limitation of the build system. Basically the capabilities of the OS are static (specified in files in `src/build-data/os`) rather than detected based on the target system.

A few options:
 a) Add build time feature detection ala autoconf. I have this on the agenda but it's a big project and I don't have a huge amount of time for botan right now.
 b) Add an option that lets a user enable features that we don't know about. Eg `./configure.py --enable-os-features=getentropy`. This would be fairly simple to implement, and useful in several different scenarios.
 c) Ignore the issue until the "stable" Linux distros catch up, circa 2019, and we can just use it knowing that we probably have a recent enough glibc. Worth mentioning that musl still doesn't implement getrandom or getentropy :(

For OpenBSD it works out because getentropy has been available there for a while, and (IIRC) OpenBSD only supports their 2 most recent releases, so basically all supported versions of OpenBSD include the syscall.

In short term (b) is probably right choice.

**Comment by lambdafu** [comment]
Thanks for the explanations, which are very helpful! In this case it's more important to have an answer than an immediate solution. As a workaround, I think concerned users of botan could simple call getentropy (or even the getrandom system call directly) during initialization.  Once it returns one byte, there would be no difference anyway.

**Comment by randombit** [comment]
With #1378 merged `getentropy` can be enabled at build time with `--enable-os-features=getentropy` so closing this.

**Comment by mouse07410** [comment]
I want to add that `getentropy()` is also available on MacOS starting with 10.12 (Sierra). But it requires adding `#include <sys/random.h>` to `src/lib/entropy/entropy_src.h` file.

Could you add it?

**Comment by mouse07410** [comment]
This is the patch I'm thinking of:
```
diff --git a/src/lib/entropy/entropy_src.h b/src/lib/entropy/entropy_src.h
index 56e5bd53e..90aacd4ce 100644
--- a/src/lib/entropy/entropy_src.h
+++ b/src/lib/entropy/entropy_src.h
@@ -10,6 +10,13 @@
 
 #include <botan/secmem.h>
 #include <botan/rng.h>
+#if defined(__APPLE__)
+#if defined(__MAC_OS_X_VERSION_MIN_REQUIRED)
+#if __MAC_OS_X_VERSION_MIN_REQUIRED >= 101200
+#include <sys/random.h>
+#endif
+#endif /* __MAC_OS_X_VERSION_MIN_REQUIRED */
+#endif /* __APPLE__ */
 #include <string>
 #include <chrono>
 #include <memory>
```

**Comment by mouse07410** [comment]
@randombit MacOS Sierra and newer include support for `getentropy()`, but it requires including `<sys/random.h>` file.

I'd like to add that include, but since it would have to be screened by the OS version (not supported prior to 10.12), I would like your guidance - what file should this include be added to?

**Comment by randombit** [comment]
@mouse07410 It should just be in `getentropy.cpp` since that's the only file that calls `getentropy`.

**Comment by mouse07410** [comment]
Please see https://github.com/randombit/botan/pull/1481
